modules/47_regipol/regiCarbonPrice/postsolve.gms

```diff
@@ -195,8 +195,13 @@ pm_emiMktTarget_dev_iter(iteration, ttot,ttot2,ext_regi,emiMktExt) = pm_emiMktTa
 loop((ext_regi,ttot2)$regiANDperiodEmiMktTarget_47(ttot2,ext_regi),
   p47_targetConverged(ttot2,ext_regi) = 1;
   loop((ttot,emiMktExt,target_type_47,emi_type_47)$((pm_emiMktTarget(ttot,ttot2,ext_regi,emiMktExt,target_type_47,emi_type_47))),
-    if((abs(pm_emiMktTarget_dev(ttot,ttot2,ext_regi,emiMktExt)) > 0.01), !! if emiMKt target did not converged
-      p47_targetConverged(ttot2,ext_regi) = 0;
+    loop(regi$regi_groupExt(ext_regi,regi),
+      loop(emiMkt$emiMktGroup(emiMktExt,emiMkt), 
+***     if emiMKt target did not converged and we are not forcing negative prices (if the price is at minimal level (1$/tCO2) and current emissions are lower than the target) 
+        if(((abs(pm_emiMktTarget_dev(ttot,ttot2,ext_regi,emiMktExt)) gt cm_emiMktTarget_tolerance) and (not ((pm_taxemiMkt(ttot2,regi,emiMkt) eq 1*sm_DptCO2_2_TDpGtC) and (pm_emiMktTarget_dev(ttot,ttot2,ext_regi,emiMktExt) lt 0)))), !! if emiMKt target did not converged and 
```

Comment:
2 questions: 

1. does GAMS store the `pm_taxemiMkt(ttot2,regi,emiMkt)` with the same number of digits so that `(not ((pm_taxemiMkt(ttot2,regi,emiMkt) eq 1*sm_DptCO2_2_TDpGtC)` will really be returned true?

2. could there be the problem that the model has in one iteration quite low emissions, uses the slope to calculate new (negative) CO2 prices that in the end miss the target, but the algorithm never checks this as it stops in the first iteration? 

maybe add a "needs to be 2 iterations in negative prices" safety buffer to prevent this from happening? or do you think this phenomenon will be very unlikely?

Reply:
1. `sm_DptCO2_2_TDpGtC` is a scalar hard-coded to `0.00366667`. So yes, this number should be always the same and equal to `0.00366667` as consequence. Tolerances only affect variables and equations as far as I know.

2. All targets are checked at every iteration if they are converging or not, so the algorithm keeps checking if the convergence has changed, even if the model reached full convergence in some previous iteration. See: https://github.com/remindmodel/remind/blob/252343cf0b39d827317e6b7affbd86a69729e37e/modules/47_regipol/regiCarbonPrice/postsolve.gms#L181C1-L208

Reply:
Still about 2, if a slope causes negative prices, the prices will be set to 1 dollar in this iteration. In the next iteration the slope will be then calculated against the 1 dollar price instead. This already adds a second attempt using a different slope to avoid the negative prices. 
I am not saying the case you mention could not happen, but I don't I saw it happening yet, at least in the runs I did so far.

Avoiding this would require a bit more complicated code than checking negative prices in two consecutive runs. I only see the need to spend time with this if this becomes a recurrent issue.

Besides, as far as I saw, negative carbon taxes are much more a symptom of ill defined targets than a result of the algorithm so far. In this case, getting the modeler attention with a non-converged run is a much better result than handing these cases quietly inside the algorithm "black-box".